backend/app/modules/multisig_signatures/service/offchain.py

```diff
@@ -69,10 +72,14 @@ def approve_pending_signatures(
 
         return approved_signatures
 
+    def _apply_atomically(self, pending_signature: MultisigSignatures):
+        with self._lock:
+            pending_signature.status = SigStatus.APPROVED
+
     def apply_staged_signatures(self, _: Context, signature_id: int):
         for idx, pending_signature in enumerate(self.staged_signatures):
             if pending_signature.id == signature_id:
-                pending_signature.status = SigStatus.APPROVED
+                self._apply_atomically(pending_signature)
                 db.session.commit()
                 self.staged_signatures.pop(idx)
                 return
```

Comment:
Disclaimer! I'm not very well versed in the way multithreading works in Python (touched it few times at most).

What I see in apply_staged_signatures is a race when two calls to this functions are happening simultaneously.

They will tread carefully when changing the status of particular entries (this change introduces a lock there), but line 84 with .pop will operate on idx which will be pointing to a wrong entry after other thread pops.

Reply:
Why not acquire the lock when entering `apply_staged_signatures` and release it when exiting?

Reply:
Just because lists in Python are thread-safe by default in CPython implementation and such methods as ```.pop()``` and ```.append()``` won't result in any kind of data race.

Reply:
And we can interate over the original staged_signatures without making any copy whilst modyfing it because we do a return just after ```.pop``` operation

Reply:
.pop is thread-safe. But the value which is stored in idx isn't. Two enter, both update idx==3 in a list of [a,b,c,d,e,f], after that one deletes d, so that list now is [a,b,c,e,f] other deletes e, since now idx == 3 points to e.

Reply:
Ok, changed